File: multiplication/src/main/java/learn/mssb/mssbmultiplication/challenge/ChallengeGeneratorServiceImpl.java
package learn.mssb.mssbmultiplication.challenge;

import org.springframework.stereotype.Service;

import java.util.Random;

@Service
public class ChallengeGeneratorServiceImpl implements ChallengeGeneratorService {

    private static final int UPPER_BOUND_FOR_RANDOM = MAXIMUM_FACTOR - MINIMUM_FACTOR + 1;

    private final Random random;

    ChallengeGeneratorServiceImpl() {
        this.random = new Random();
    }

    protected ChallengeGeneratorServiceImpl(Random random) {
        this.random = random;
    }

    @Override
    public Challenge randomChallenge() {
        return new Challenge(nextFactor(), nextFactor());
    }

    private int nextFactor() {
        return MINIMUM_FACTOR + random.nextInt(UPPER_BOUND_FOR_RANDOM);
    }
}

File: multiplicationkt/src/main/kotlin/challenge/ChallengeGeneratorServiceImpl.kt
package learn.mssbkt.multiplicationkt.challenge

import org.springframework.stereotype.Service
import kotlin.random.Random

private const val MINIMUM_FACTOR = 11
private const val MAXIMUM_FACTOR = 100
private const val MIN_TO_MAX_RANGE = MAXIMUM_FACTOR - MINIMUM_FACTOR

@Service
class ChallengeGeneratorServiceImpl(
    private val random: Random = Random.Default
) : ChallengeGeneratorService {
    override fun randomChallenge(): Challenge {
        return Challenge(factorA = next(), factorB = next())
    }

    private fun next(): Int = random.nextInt(MIN_TO_MAX_RANGE) + MINIMUM_FACTOR
}
